File: 310971/Stankevich/jpa320/src/main/java/com/task/rest/service/AuthorService.java
package com.task.rest.service;

import com.task.rest.dto.AuthorRequestTo;
import com.task.rest.dto.AuthorResponseTo;
import com.task.rest.exception.ResourceNotFoundException;
import com.task.rest.model.Author;
import com.task.rest.repository.AuthorRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class AuthorService {

    private final AuthorRepository authorRepository;

    @Transactional(readOnly = true)
    public AuthorResponseTo getById(Long id) {
        log.info("Getting author by id: {}", id);
        Author author = authorRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Author not found with id: {}", id);
                    return new ResourceNotFoundException("Author not found with id: " + id);
                });
        return mapToResponse(author);
    }

    @Transactional(readOnly = true)
    public Page<AuthorResponseTo> getAll(Pageable pageable) {
        log.info("Getting all authors with pageable: {}", pageable);
        return authorRepository.findAll(pageable)
                .map(this::mapToResponse);
    }

    public AuthorResponseTo create(AuthorRequestTo requestTo) {
        log.info("Creating new author with login: {}", requestTo.getLogin());

        if (authorRepository.existsByLogin(requestTo.getLogin())) {
            log.error("Author with login already exists: {}", requestTo.getLogin());
            throw new ResponseStatusException(HttpStatus.FORBIDDEN,
                    "Author with login already exists: " + requestTo.getLogin());
        }

        Author author = new Author();
        author.setLogin(requestTo.getLogin());
        author.setPassword(requestTo.getPassword());
        author.setFirstname(requestTo.getFirstname());
        author.setLastname(requestTo.getLastname());

        author = authorRepository.save(author);
        log.info("Author created with id: {}", author.getId());

        return mapToResponse(author);
    }

    public AuthorResponseTo update(Long id, AuthorRequestTo requestTo) {
        log.info("Updating author with id: {}", id);

        Author author = authorRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Author not found with id: {}", id);
                    return new ResourceNotFoundException("Author not found with id: " + id);
                });

        authorRepository.findByLogin(requestTo.getLogin()).ifPresent(existingAuthor -> {
            if (!existingAuthor.getId().equals(id)) {
                log.error("Author with login already exists: {}", requestTo.getLogin());
                throw new ResponseStatusException(HttpStatus.FORBIDDEN,
                        "Author with login already exists: " + requestTo.getLogin());
            }
        });

        author.setLogin(requestTo.getLogin());
        author.setPassword(requestTo.getPassword());
        author.setFirstname(requestTo.getFirstname());
        author.setLastname(requestTo.getLastname());

        author = authorRepository.save(author);
        log.info("Author updated with id: {}", id);

        return mapToResponse(author);
    }

    public void delete(Long id) {
        log.info("Deleting author with id: {}", id);
        if (!authorRepository.existsById(id)) {
            log.error("Author not found with id: {}", id);
            throw new ResourceNotFoundException("Author not found with id: " + id);
        }
        authorRepository.deleteById(id);
        log.info("Author deleted with id: {}", id);
    }

    private AuthorResponseTo mapToResponse(Author author) {
        return new AuthorResponseTo(
                author.getId(),
                author.getLogin(),
                author.getFirstname(),
                author.getLastname()
        );
    }
}

File: 351002/Melnikov/publisher/services/author_service.py
from ..common.errors import ConflictError, NotFoundError
from ..common.validation import validate_length
from ..dto.author_dto import AuthorRequestTo, AuthorResponseTo
from ..models.author import Author
from ..repositories.registry import author_repository


class AuthorService:
    def create(self, dto: AuthorRequestTo) -> AuthorResponseTo:
        login = validate_length("login", dto.login, 2, 64, 1)
        password = validate_length("password", dto.password, 8, 128, 2)
        firstname = validate_length("firstname", dto.firstname, 2, 64, 3)
        lastname = validate_length("lastname", dto.lastname, 2, 64, 4)

        existing = [a for a in author_repository.get_all() if a.login == login]
        if existing:
            raise ConflictError("Author with this login already exists", 1)

        author = Author(
            id=0,
            login=login,
            password=password,
            firstname=firstname,
            lastname=lastname
        )
        created = author_repository.create(author)
        return self._to_response(created)

    def get_all(self) -> list[AuthorResponseTo]:
        return [self._to_response(author) for author in author_repository.get_all()]

    def get_by_id(self, author_id: int) -> AuthorResponseTo:
        author = author_repository.get_by_id(author_id)
        if not author:
            raise NotFoundError("Author not found", 1)
        return self._to_response(author)

    def update(self, author_id: int, dto: AuthorRequestTo) -> AuthorResponseTo:
        author = author_repository.get_by_id(author_id)
        if not author:
            raise NotFoundError("Author not found", 1)

        login = validate_length("login", dto.login, 2, 64, 1)
        password = validate_length("password", dto.password, 8, 128, 2)
        firstname = validate_length("firstname", dto.firstname, 2, 64, 3)
        lastname = validate_length("lastname", dto.lastname, 2, 64, 4)

        for item in author_repository.get_all():
            if item.login == login and item.id != author_id:
                raise ConflictError("Author with this login already exists", 1)

        author.login = login
        author.password = password
        author.firstname = firstname
        author.lastname = lastname

        updated = author_repository.update(author)
        return self._to_response(updated)

    def delete(self, author_id: int) -> None:
        deleted = author_repository.delete(author_id)
        if not deleted:
            raise NotFoundError("Author not found", 1)

    @staticmethod
    def _to_response(author: Author) -> AuthorResponseTo:
        return AuthorResponseTo(
            id=author.id,
            login=author.login,
            password=author.password,
            firstname=author.firstname,
            lastname=author.lastname
        )